unittests/Expr/ExprTest.cpp

```diff
@@ -113,4 +113,120 @@ TEST(ExprTest, ExtractConcat) {
   EXPECT_EQ(Expr::Extract, concat2->getKid(1)->getKind());
 }
 
+TEST(ExprTest, ReadExprFoldingBasic) {
+  unsigned size = 10;
+
+  // Constant array
+  std::vector<ref<ConstantExpr> > Contents(size);
+  for (unsigned i = 0; i < size; ++i)
+    Contents[i] = ConstantExpr::create(i + 1, Expr::Int8);
+  ArrayCache ac;
+  const Array *array =
+      ac.CreateArray("arr", size, &Contents[0], &Contents[0] + size);
+
+  // Basic constant folding rule
+  // Constant index (0)
+  ref<Expr> index = ConstantExpr::create(0, Expr::Int32);
+  UpdateList ul(array, 0);
+  ref<Expr> read = ReadExpr::create(ul, index);
+  // Read - should be constant folded to Contents[0]
+  // Check that constant folding worked
+  EXPECT_EQ(Expr::Constant, read.get()->getKind());
+  ASSERT_TRUE(EqExpr::create(Contents[0], read).get()->isTrue());
```

Comment:
It would not exercise different behaviour of the **current** implementation. If the implementation was later changed this might not hold so given what I've suggested is a property that we always expect to hold I suggest that we check for it.